src/edu/csus/ecs/pc2/ui/RunsTablePane.java

```diff
@@ -2196,8 +2279,13 @@ private void showSourceForSelectedRun() {
         } catch (Exception e) {
             getController().getLog().log(Log.WARNING, "Exception logged ", e);
             showMessage("Unable to show run source, check log");
+            
+            //make sure this is clear in case of exception
+            requestedRun = null;
         }
-
+        
+        // OK to now start another view source now
+        AllowViewSource();
```

Comment:
This is a followup comment to the one I made above, around (new) line 2198.  Suppose a user clicks the `View Source` button.  That triggers a call to `BlockViewSource()`, followed by a spawning of a Thread (call it Thread1) whose `run()` method does just one thing:  calls method `showSourceForSelectedRun()`.  At some point during the execution of `showSourceForSelectedRun()` on Thread1, the method calls `AllowViewSource()` (line 2198, above).  

Now suppose a user clicks the `View Source` button again, _after_ the execution of the first  `AllowViewSource()` call in Thread1 but _before_ Thread1 completes the execution of `showSourceForSelectedRun()`.  This second button click will call `BlockViewSource()`, and then will spawn a new Thread (call it Thread2) which runs a separate copy of method `showSourceForSelectedRun()`.   So Thread2 starts executing in `showSourceForSelectedRun()`.

Now suppose Thread1 finishes -- the last statement being a call to `AllowViewSource()` (the line to which this comment is attached).  We now have the situation that Thread2 is executing within method `showSourceForSelectedRun()` with the `showSourceActive` variable set to false -- but that's wrong; there **_IS_** an active "showSource" operation in progress.

I'm not positive, but I think that removing the FIRST call to `AllowShowSource()`, leaving just the one at the end of the method, might fix the problem.

Or maybe I'm making this whole thing up in my head?

Reply:
Yes, it's actually missing a 'return' at line 2272.  Removing the first AllowShowSource() (line 2198) would break the 2 checks that immediately follow since they could 'return' without doing an AllowViewSource():
```
                // OK to now start another view source
                AllowViewSource();
               
                // if any type of error occurred requesting the run to view, log a summary message and finish
                if(bFetchError) {
                    getController().getLog().log(Log.WARNING, "Unable to fetch run " + run.getNumber() + " at site " + run.getSiteNumber() + " from server");
                    showMessage("Unable to fetch selected run; check log");
                   return;
                }
                
                //if we get here we know there should be RunFiles in the contest model -- but let's sanity-check that
                if (!getContest().isRunFilesPresent(run)) {
                    
                    //something bad happened -- we SHOULD have RunFiles at this point!
                    getController().getLog().log(Log.SEVERE, "Unable to find RunFiles for run " + run.getNumber() + " at site " + run.getSiteNumber() + " -- server error?");
                    showMessage("Unable to fetch selected run; check log");
                    return;

                } else {
```

What may be better is to do something like:

```
                // OK to now start another view source
                AllowViewSource();
               
                // if any type of error occurred requesting the run to view, log a summary message and finish
                if(bFetchError) {
                    getController().getLog().log(Log.WARNING, "Unable to fetch run " + run.getNumber() + " at site " + run.getSiteNumber() + " from server");
                    showMessage("Unable to fetch selected run; check log");
                } else {
           
                    //if we get here we know there should be RunFiles in the contest model -- but let's sanity-check that
                    if (!getContest().isRunFilesPresent(run)) {
                    
                        //something bad happened -- we SHOULD have RunFiles at this point!
                        getController().getLog().log(Log.SEVERE, "Unable to find RunFiles for run " + run.getNumber() + " at site " + run.getSiteNumber() + " -- server error?");
                        showMessage("Unable to fetch selected run; check log");

                    } else {
                        //get the Run files
                        ...
                   }
              }
             return;   // THIS IS THE NEW return STATEMENT
         } else {
                // getContest().getRun() returned null
                getController().getLog().log(Log.WARNING, "Selected run not found");
                showMessage("Selected run not found");
         }

    

```

Reply:
Ok, adding that `return` does seem like it would fix the problem.  Can you push an update to your fork so I can pull it down for testing?

Reply:
Fixed comment describing possible NPE (which, by the I caused to happen on purpose).
Fix missing return and optimized code flow.